clash-lib/src/Clash/Core/HasType.hs

```diff
@@ -0,0 +1,222 @@
+import Clash.Util (pprPanic)
+import qualified Clash.Util.Interpolate as I
+
+class HasType a where
+  coreTypeOf :: a -> Type
+
+instance HasType DataCon where
+  coreTypeOf = dcType
+
+instance HasType Literal where
+  coreTypeOf = \case
+    IntegerLiteral _ -> integerPrimTy
+    IntLiteral _ -> intPrimTy
+    WordLiteral _ -> wordPrimTy
+    StringLiteral _ -> stringPrimTy
+    FloatLiteral _ -> floatPrimTy
+    DoubleLiteral _ -> doublePrimTy
+    CharLiteral _ -> charPrimTy
+    Int64Literal _ -> int64PrimTy
+    Word64Literal _ -> word64PrimTy
+    NaturalLiteral _ -> naturalPrimTy
+    ByteArrayLiteral _ -> byteArrayPrimTy
+
+instance HasType PrimInfo where
+  coreTypeOf pr =
+    case primMultiResult pr of
+      SingleResult -> primType pr
+
+      -- See Note [MultiResult type] in Clash.Normalize.Transformations.MultiPrim
+      MultiResult
+        | let (primArgs, primResTy) = splitFunForallTy (primType pr)
+        , TyConApp tupTcNm tupArgs <- tyView primResTy
+        , Text.isInfixOf "GHC.Tuple.(" (nameOcc tupTcNm)
+        -> mkPolyFunTy primResTy (primArgs <> fmap Right tupArgs)
+
+        | otherwise
+        -> error "PrimInfo.coreTypeOf: MultiResult primitive without tuple type"
+
+instance HasType Type where
+  coreTypeOf = id
+
+instance HasType (Var a) where
+  coreTypeOf = varType
+
+class InferType a where
+  inferCoreTypeOf :: TyConMap -> a -> Type
+
+instance InferType Term where
+  inferCoreTypeOf tcm = go
+   where
+    go = \case
+      Var i -> coreTypeOf i
+      Data dc -> coreTypeOf dc
+      Literal l -> coreTypeOf l
+      Prim pr -> coreTypeOf pr
+      Lam i x -> mkFunTy (coreTypeOf i) (go x)
+      TyLam i x -> ForAllTy i (go x)
+
+      x@App{} ->
+        case collectArgs x of
+          (fun, args) -> applyTypeToArgs x tcm (go fun) args
+
+      x@TyApp{} ->
+        case collectArgs x of
+          (fun, args) -> applyTypeToArgs x tcm (go fun) args
+
+      Letrec _ x -> go x
+      Case _ _ [] -> undefinedTy
+      Case _ _ (x:_) -> go (snd x)
+      Cast _ _ a -> a
+      Tick _ x -> go x
+
+-- | Get the result type of a polymorphic function given a list of arguments
+applyTypeToArgs
+  :: Term
+  -> TyConMap
+  -> Type
+  -> [Either Term Type]
+  -> Type
+applyTypeToArgs e m opTy args = go opTy args
+ where
+  go opTy' []               = opTy'
+  go opTy' (Right ty:args') = goTyArgs opTy' [ty] args'
+  go opTy' (Left a:args')   = case splitFunTy m opTy' of
+    Just (_,resTy) -> go resTy args'
+    _ -> error [I.i|
+        Unexpected application. The term
+
+          #{showPpr e}
+
+        applied an argument
+
+          #{showPpr a}
+
+        to something with the non-function type
+
+          #{showPpr opTy'}
+      |]
+
+  goTyArgs opTy' revTys (Right ty:args') = goTyArgs opTy' (ty:revTys) args'
+  goTyArgs opTy' revTys args'            = go (piResultTys m opTy' (reverse revTys)) args'
+
+-- | Like 'piResultTyMaybe', but errors out when a type application is not
```

Comment:
`piResultTyMaybe` was deleted?

Reply:
It was only used to implement `piResultTy`, which I removed (because I didn't see any noticeable change in perf by changing `piResultTy` to be defined in terms of `piResultTys`)

Reply:
Oh, but the comment!